src/rules/no-self-destination.ts

```diff
@@ -0,0 +1,103 @@
+import { createRule } from "../utils/index.ts";
+import type { Link, Image, Definition } from "mdast";
+import path from "node:path";
+
+export default createRule("no-self-destination", {
+    meta: {
+        type: "problem",
+        docs: {
+            description: "disallow redundant self-destination links",
+            categories: ["recommended"],
+            listCategory: "Markdown Link",
+        },
+        fixable: undefined,
+        hasSuggestions: false,
+        schema: [],
+        messages: {
+            redundantSelfDestination: "Redundant self-destination link '{{url}}'. Use '{{suggestion}}' instead.",
+        },
+    },
+    create(context) {
+        const filename = path.isAbsolute(context.physicalFilename)
+            ? context.physicalFilename
+            : path.resolve(context.cwd, context.physicalFilename);
+        const basename = path.basename(filename);
+        
+        /**
+         * Checks if a URL refers to the current file
+         */
+        function isSelfDestination(url: string): boolean {
+            // Skip external URLs, fragments only, and empty URLs
+            if (!url || url.startsWith('http://') || url.startsWith('https://') || url.startsWith('#') || url.startsWith('//')) {
```

Comment:
Things like mailto protocols should also be excluded.

```suggestion
            if (!url || /^(?:[a-z]+:|\/\/)/iu.test(url) || url.startsWith('#')) {
```